### PR TITLE
Adding Sequential Support for BLIF reader and writer

### PR DESCRIPTION
**Pull request by maustin8**
Hey,

I've added some sequential support to the BLIF IO. I have updated the way that latch information is stored so that the full 5-argument latch declaration (input, output, type, control, initial value) is supported and this information can be carried through to the writers. The BLIF writer is the only writer that has been updated in this PR to write out this information. BLIF files written by ABC and Yosys have been used to test this support. The tests included with Mockturtle were used to verify that no other features were broken with this addition.

### CODE REVIEW

**Comment by hriener** [comment]
Thank you for your contribution. The PR looks pretty good; I need a bit more time to completely review the code and will come back to you soon.

**Comment by hriener** [comment]
Closed because of inactivity.

We can reopen the PR when there is time to work on the requested tests.

**Comment by maustin8** [comment]
I have added a test for the BLIF reader and the 'split' function change into test/io/blif_reader.cpp. Is this sufficient?

**Comment by hriener** [comment]
Have a look at the output of coveralls:
https://coveralls.io/builds/28773288/source?filename=include/lorina/blif.hpp

Two test cases would be helpful:
- one unit test for `.latch`
- one regression test for the changed `split` function

**Comment by maustin8** [comment]
Those test cases are covered in tests/io/blif_reader.cpp and I'm not sure why the .latch section is not reached in blif.hpp in the build number that you have linked but this seems to be resolved in build 208 of pull #28 where coverage has increased 0.4%. If I am mistaken, let me know.

**Comment by hriener** [comment]
> Those test cases are covered in tests/io/blif_reader.cpp and I'm not sure why the .latch section is not reached in blif.hpp in the build number that you have linked but this seems to be resolved in build 208 of pull #28 where coverage has increased 0.4%. If I am mistaken, let me know.

Yes, coverage is good (~88%) with the two new test cases. However, I also slightly changed the implementation and API in `lorina`. Please have a look and check. Then, we can merge it!

File: include/mockturtle/io/write_blif.hpp [hunk 1]
```diff
@@ -85,95 +90,139 @@ void write_blif( Ntk const& ntk, std::ostream& os )
   static_assert( has_node_to_index_v<Ntk>, "Ntk does not implement the node_to_index method" );
   static_assert( has_node_function_v<Ntk>, "Ntk does not implement the node_function method" );
 
-  topo_view topo_ntk{ntk};
+  // topo_view topo_ntk{ntk};
```
Comment (hriener) [inline include/mockturtle/io/write_blif.hpp]: I am not completely sure about this line. Why?

File: include/mockturtle/io/write_blif.hpp [hunk 2]
```diff
@@ -85,95 +90,139 @@ void write_blif( Ntk const& ntk, std::ostream& os )
...
   {
       os << ".inputs ";
-      topo_ntk.foreach_pi( [&]( auto const& n ) {
+      topo_ntk.foreach_ci( [&]( auto const& n, auto index ) {
+        if ( ( ( index + 1 ) <= topo_ntk.num_cis() - topo_ntk.num_latches() ) ) {
           if constexpr ( has_has_name_v<Ntk> && has_get_name_v<Ntk> )
           {
             signal<Ntk> const s = topo_ntk.make_signal( topo_ntk.node_to_index( n ) );
-            std::string const name = topo_ntk.has_name( s ) ? topo_ntk.get_name( s ) : fmt::format( "c_{}", topo_ntk.get_node( s ) );
+            std::string const name = topo_ntk.has_name( s ) ? topo_ntk.get_name( s ) : fmt::format( "pi{}", topo_ntk.get_node( s ) );
```
Comment (hriener) [inline include/mockturtle/io/write_blif.hpp]: Do you mean "new_n{}" instead of "pi{}"?

File: include/mockturtle/io/write_blif.hpp [hunk 3]
```diff
@@ -85,95 +90,139 @@ void write_blif( Ntk const& ntk, std::ostream& os )
...
             os << name << ' ';
           }
           else
           {
-            os << fmt::format( "c_n{} ", topo_ntk.node_to_index( n ) );
+            os << fmt::format( "pi{} ", topo_ntk.node_to_index( n ) );
```
Comment (hriener) [inline include/mockturtle/io/write_blif.hpp]: Do you mean "new_n{}" instead of "pi{}"?

File: include/mockturtle/io/write_blif.hpp (hunk 1 above)
Comment (maustin8) [reply include/mockturtle/io/write_blif.hpp]: I undid this change so networks do not need to be placed in topological order before calling write_blif.

File: include/mockturtle/io/write_blif.hpp (hunk 2 above)
Comment (maustin8) [reply include/mockturtle/io/write_blif.hpp]: No. Like ABC, I made the default for the PIs that do not have an originally defined name pi{index}. I made new_n{} the default just for internal nodes.

File: include/mockturtle/io/write_blif.hpp (hunk 3 above)
Comment (maustin8) [reply include/mockturtle/io/write_blif.hpp]: Same reason as the comment above.

**Comment by hriener** [comment]
Thanks. In my opinion, the chances look good. However, the test cases do not pass anymore:
https://github.com/maustin8/mockturtle/actions

Could you double-check if the simulation values are correct?

**Comment by maustin8** [comment]
I fixed the test file for the BLIF reader. I checked the simulation values against ABC and also performed the truth table operations by hand to get the new values.

**Comment by hriener** [comment]
Great! Thank you for your contribution.

**Comment by lee30sonia** [comment]
To whoever that has an opinion, which could also be some external users who were not involved before but use these interfaces.

**Comment by hriener** [comment]
Well, #61 is from 2018 and #245 from 2020. I believe most people involved have moved on. I would suggest opening a new issue on BLIF with white and black boxes. Also, it would make sense to add a reference to the black/white box proposal if it exists. Has that been introduced in `Yosys`/`ABC`?

In 2018, I implemented a BDD-based model checker on top of `mockturtle` and `Sylvan`. I needed to read some benchmarks and for that I quickly implemented the sequential interface. The model checker is not part of `mockturtle`.

Max Austin and Walter Lau Neto started to work on sequential synthesis. I am not aware of the results; I just remember that Walter tried to integrate sequential cuts [1] into `rewriting`.

[1] http://people.eecs.berkeley.edu/~alanmi/publications/2007/iwls07_seq.pdf

**Pull request merged**